reader.go

```diff
@@ -0,0 +1,160 @@
+package gopc
+
+import (
+	"io"
+	"path/filepath"
+	"strings"
+)
+
+type archiveFile interface {
+	Open() (io.ReadCloser, error)
+	Name() string
+}
+
+type archive interface {
+	Files() []archiveFile
+}
+
+// Reader implements a OPC file reader.
+type Reader struct {
+	Parts         []*Part
+	Relationships []*Relationship
+	Properties    CoreProperties
+	p             *pkg
+	r             archive
+}
+
+// NewReader returns a new Reader reading an OPC file to r.
+func NewReader(r io.ReaderAt, size int64) (*Reader, error) {
+	zr, err := newZipReader(r, size)
+	if err != nil {
+		return nil, err
+	}
+	return newReader(zr)
+}
+
+// newReader returns a new Reader reading an OPC file to r.
+func newReader(a archive) (*Reader, error) {
+	r := &Reader{p: newPackage(), r: a}
+	if err := r.loadPackage(); err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func (r *Reader) loadPackage() error {
+	files := r.r.Files()
+	ct, rels, err := r.loadPartPropierties()
+	if err != nil {
+		return err
+	}
+	r.Parts = make([]*Part, 0, len(files)-1) // -1 is for [Content_Types].xml
+	for _, file := range files {
+		fileName := "/" + file.Name()
+		// skip content types part, relationship parts and directories
+		if strings.EqualFold(fileName, contentTypesName) || isRelationshipURI(fileName) || strings.HasSuffix(fileName, "/") {
+			continue
+		}
+		if strings.EqualFold(fileName, r.Properties.PartName) {
+			cp, err := r.loadCoreProperties(file)
+			if err != nil {
+				return err
+			}
+			r.Properties = *cp
+			continue
+		}
+		cType, err := ct.findType(fileName)
+		if err != nil {
+			return err
+		}
+		part := &Part{Name: fileName, ContentType: cType, Relationships: rels.findRelationship(fileName)}
+		r.Parts = append(r.Parts, part)
+		r.p.add(part)
+	}
+	r.p.contentTypes = *ct
+	return nil
+}
+
+func (r *Reader) loadPartPropierties() (*contentTypes, *relationshipsPart, error) {
+	var ct *contentTypes
+	rels := new(relationshipsPart)
+	var err error
+	files := r.r.Files()
+	for _, file := range files {
+		name := "/" + file.Name()
+		if strings.EqualFold(name, contentTypesName) {
+			ct, err = r.loadContentType(file)
+			if err != nil {
```

Comment:
move err check after the if/elseif to unify all the following error checks

Reply:
moved to the end of the for